framework/physics/TimeStepControllers/TimeStepper.h

```diff
@@ -0,0 +1,85 @@
+#ifndef CHITECH_TIMESTEPPER_H
+#define CHITECH_TIMESTEPPER_H
+
+#include "ChiObject.h"
+
+namespace chi_physics
+{
+
+enum class TimeStepStatus
+{
+  SUCCESS = 0,
+  FAILURE = 1,
+  NEUTRAL = 2
+};
+
+/**Base class for all timestep controllers.*/
+class TimeStepper : public ChiObject
+{
+public:
+  /**Overridable method to get the timestep size.*/
+  double TimeStepSize() const;
+
+  /**Returns the current controller time.*/
+  double Time() const;
+
+  /**Returns the current time index.*/
+  size_t TimeStepIndex() const;
+
+  /**Returns the current controller start_time.*/
+  double StartTime() const;
+
+  /**Returns the current controller end_time.*/
+  double EndTime() const;
+
+  /**Returns the current controller max time steps.*/
+  double MaxTimeSteps() const;
+
+  /**If start_time <= time <= end_time, this will return true.*/
+  bool IsActive() const;
+
+
+  /**Manually set the time step size.*/
+  void SetTimeStepSize(double dt);
+
+  /**Manually set the current time.*/
+  void SetTime(double time);
+
+  /**Manually set the start_time.*/
+  void SetStartTime(double time);
+
+  /**Manually set the end_time.*/
+  void SetEndTime(double time);
+
+  /**Manually set the maximum number of time steps. A negative number disables
+* this check.*/
+  void SetMaxTimeSteps(int n);
+
+  /**Advances the controller's state. The most basic action here is to
+  * advance time and the time index.*/
+  virtual void Advance();
+
+  /**Adapts according to the timestep status.*/
+  virtual bool Adapt(TimeStepStatus time_step_status) {return false;}
```

Comment:
Is this a function to return whether or not to adapt the time step? The function name and doc doesn't scream that to me, but rather that the function computes the adaptive time step. Maybe `DoAdapt` would be better with doc that says `Returns whether time step adaption should be performed.`.

Reply:
Changed to:
```
/**Adapts according to the timestep status. If it could provide a change
  * it returns true, otherwise false.*/
  virtual bool Adapt(TimeStepStatus time_step_status) {return false;}
```